Build error on kernel 4.13.x

Hi, i got a compile error when building the module please help. I really like this project and would love to get it working!

My dependencies:
```
sudo dnf install dkms @development-tools "kernel-devel-uname-r == $(uname -r)"
```

Make:
```
# make dkms                                                                                                        
Installing Douane Linux kernel module version 0.8.2...                

Creating symlink /var/lib/dkms/douane/0.8.2/source ->                 
                 /usr/src/douane-0.8.2                                

DKMS: add completed.               

Kernel preparation unnecessary for this kernel.  Skipping...          

Building module:                   
cleaning build area...             
make -j4 KERNELRELEASE=4.13.12-300.fc27.x86_64 -C /lib/modules/4.13.12-300.fc27.x86_64/build M=/var/lib/dkms/douane/0.8.2/build...(bad exit status: 2)
Error! Bad return status for module build on kernel: 4.13.12-300.fc27.x86_64 (x86_64)                                                       
Consult /var/lib/dkms/douane/0.8.2/build/make.log for more information.                                                                     
make: *** [Makefile:86: dkms] Error 10
```

Logfile:
```
# cat /var/lib/dkms/douane/0.8.2/build/make.log
DKMS make.log for douane-0.8.2 for kernel 4.13.12-300.fc27.x86_64 (x86_64)
Sun Nov 19 11:08:59 CET 2017
make[1]: Entering directory '/usr/src/kernels/4.13.12-300.fc27.x86_64'
  AR      /var/lib/dkms/douane/0.8.2/build/built-in.o
  CC [M]  /var/lib/dkms/douane/0.8.2/build/douane.o
/var/lib/dkms/douane/0.8.2/build/douane.c: In function ‘task_exe_path’:
/var/lib/dkms/douane/0.8.2/build/douane.c:713:5: error: implicit declaration of function ‘task_lock’; did you mean ‘txd_lock’? [-Werror=implicit-function-declaration]
     task_lock(locked_task);
     ^~~~~~~~~
     txd_lock
/var/lib/dkms/douane/0.8.2/build/douane.c:734:11: error: implicit declaration of function ‘task_unlock’; did you mean ‘tasklet_unlock’? [-Werror=implicit-function-declaration]
           task_unlock(locked_task);
           ^~~~~~~~~~~
           tasklet_unlock
/var/lib/dkms/douane/0.8.2/build/douane.c:759:5: error: expected ‘while’ before ‘while_each_thread’
   } while_each_thread(task, locked_task);
     ^~~~~~~~~~~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c: In function ‘find_task_from_socket_file’:
/var/lib/dkms/douane/0.8.2/build/douane.c:773:3: error: implicit declaration of function ‘for_each_process’; did you mean ‘for_each_node’? [-Werror=implicit-function-declaration]
   for_each_process(task)
   ^~~~~~~~~~~~~~~~
   for_each_node
/var/lib/dkms/douane/0.8.2/build/douane.c:774:3: error: expected ‘;’ before ‘{’ token
   {
   ^
/var/lib/dkms/douane/0.8.2/build/douane.c: In function ‘initialize_module’:
/var/lib/dkms/douane/0.8.2/build/douane.c:1196:3: error: implicit declaration of function ‘nf_register_hook’; did you mean ‘nf_register_net_hook’? [-Werror=implicit-function-declaration]
   nf_register_hook(&nfho_outgoing);
   ^~~~~~~~~~~~~~~~
   nf_register_net_hook
/var/lib/dkms/douane/0.8.2/build/douane.c: In function ‘exit_module’:
/var/lib/dkms/douane/0.8.2/build/douane.c:1224:3: error: implicit declaration of function ‘nf_unregister_hook’; did you mean ‘nf_unregister_net_hook’? [-Werror=implicit-function-declaration]
   nf_unregister_hook(&nfho_outgoing);
   ^~~~~~~~~~~~~~~~~~
   nf_unregister_net_hook
cc1: some warnings being treated as errors
make[2]: *** [scripts/Makefile.build:309: /var/lib/dkms/douane/0.8.2/build/douane.o] Error 1
make[1]: *** [Makefile:1516: _module_/var/lib/dkms/douane/0.8.2/build] Error 2
make[1]: Leaving directory '/usr/src/kernels/4.13.12-300.fc27.x86_64'
```

any help appreciated :)

Hi @zedtux , thanks alot for your answer.

I tried to change the functions but without success. I have to say i'm not very fluent with C and i never did code modules or something like that. 

As far as i can tell the problem is now with the struct nf_hook_ops which needs additional parameters.
`struct nf_hook_ops {
	/* User fills in from here down. */
	nf_hookfn		*hook;
	struct net_device	*dev;
	void			*priv;
	u_int8_t		pf;
	unsigned int		hooknum;
	/* Hooks are ordered in ascending priority. */
	int			priority;
};`

`
DKMS make.log for douane-0.8.2 for kernel 4.13.12-300.fc27.x86_64 (x86_64)
Tue Nov 21 10:28:39 CET 2017
make[2]: Entering directory '/usr/src/kernels/4.13.12-300.fc27.x86_64'
  AR      /var/lib/dkms/douane/0.8.2/build/built-in.o
  CC [M]  /var/lib/dkms/douane/0.8.2/build/douane.o
/var/lib/dkms/douane/0.8.2/build/douane.c: In function ‘task_exe_path’:
/var/lib/dkms/douane/0.8.2/build/douane.c:713:5: error: implicit declaration of function ‘task_lock’; did you mean ‘txd_lock’? [-Werror=implicit-function-declaration]
     task_lock(locked_task);
     ^~~~~~~~~
     txd_lock
/var/lib/dkms/douane/0.8.2/build/douane.c:734:11: error: implicit declaration of function ‘task_unlock’; did you mean ‘tasklet_unlock’? [-Werror=implicit-function-declaration]
           task_unlock(locked_task);
           ^~~~~~~~~~~
           tasklet_unlock
/var/lib/dkms/douane/0.8.2/build/douane.c:759:5: error: expected ‘while’ before ‘while_each_thread’
   } while_each_thread(task, locked_task);
     ^~~~~~~~~~~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c: In function ‘find_task_from_socket_file’:
/var/lib/dkms/douane/0.8.2/build/douane.c:773:3: error: implicit declaration of function ‘for_each_process’; did you mean ‘for_each_node’? [-Werror=implicit-function-declaration]
   for_each_process(task)
   ^~~~~~~~~~~~~~~~
   for_each_node
/var/lib/dkms/douane/0.8.2/build/douane.c:774:3: error: expected ‘;’ before ‘{’ token
   {
   ^
/var/lib/dkms/douane/0.8.2/build/douane.c: In function ‘initialize_module’:
/var/lib/dkms/douane/0.8.2/build/douane.c:1196:24: error: passing argument 1 of ‘nf_register_net_hook’ from incompatible pointer type [-Werror=incompatible-pointer-types]
   nf_register_net_hook(&nfho_outgoing);
                        ^
In file included from /var/lib/dkms/douane/0.8.2/build/douane.c:20:0:
./include/linux/netfilter.h:154:5: note: expected ‘struct net *’ but argument is of type ‘struct nf_hook_ops *’
 int nf_register_net_hook(struct net *net, const struct nf_hook_ops *ops);
     ^~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c:1196:3: error: too few arguments to function ‘nf_register_net_hook’
   nf_register_net_hook(&nfho_outgoing);
   ^~~~~~~~~~~~~~~~~~~~
In file included from /var/lib/dkms/douane/0.8.2/build/douane.c:20:0:
./include/linux/netfilter.h:154:5: note: declared here
 int nf_register_net_hook(struct net *net, const struct nf_hook_ops *ops);
     ^~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c: In function ‘exit_module’:
/var/lib/dkms/douane/0.8.2/build/douane.c:1224:26: error: passing argument 1 of ‘nf_unregister_net_hook’ from incompatible pointer type [-Werror=incompatible-pointer-types]
   nf_unregister_net_hook(&nfho_outgoing);
In file included from /var/lib/dkms/douane/0.8.2/build/douane.c:20:0:
./include/linux/netfilter.h:155:6: note: expected ‘struct net *’ but argument is of type ‘struct nf_hook_ops *’
 void nf_unregister_net_hook(struct net *net, const struct nf_hook_ops *ops);
      ^~~~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c:1224:3: error: too few arguments to function ‘nf_unregister_net_hook’
   nf_unregister_net_hook(&nfho_outgoing);
   ^~~~~~~~~~~~~~~~~~~~~~
In file included from /var/lib/dkms/douane/0.8.2/build/douane.c:20:0:
./include/linux/netfilter.h:155:6: note: declared here
 void nf_unregister_net_hook(struct net *net, const struct nf_hook_ops *ops);
      ^~~~~~~~~~~~~~~~~~~~~~
cc1: some warnings being treated as errors
make[3]: *** [scripts/Makefile.build:309: /var/lib/dkms/douane/0.8.2/build/douane.o] Error 1
make[2]: *** [Makefile:1516: _module_/var/lib/dkms/douane/0.8.2/build] Error 2
make[2]: Leaving directory '/usr/src/kernels/4.13.12-300.fc27.x86_64'
`

If i get it working I sure will make a PR :)

@zedtux hello, thanks so I am continuing here. I applied modifications that you and @cynay  post above. But I have still problems my log: Many thanks fro help. 
```

DKMS make.log for douane-0.8.2 for kernel 4.14.0-kali1-amd64 (x86_64)
Ut jan  9 16:31:31 CET 2018
make[1]: Entering directory '/usr/src/linux-headers-4.14.0-kali1-amd64'
  AR      /var/lib/dkms/douane/0.8.2/build/built-in.o
  CC [M]  /var/lib/dkms/douane/0.8.2/build/douane.o
/var/lib/dkms/douane/0.8.2/build/douane.c:1174:15: error: variable ‘nfho_outgoing’ has initializer but incomplete type
 static struct nf_net_hook_ops nfho_outgoing = {
               ^~~~~~~~~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c:1175:4: error: ‘struct nf_net_hook_ops’ has no member named ‘hook’
   .hook     = netfiler_packet_hook,
    ^~~~
/var/lib/dkms/douane/0.8.2/build/douane.c:1175:15: warning: excess elements in struct initializer
   .hook     = netfiler_packet_hook,
               ^~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c:1175:15: note: (near initialization for ‘nfho_outgoing’)
/var/lib/dkms/douane/0.8.2/build/douane.c:1176:4: error: ‘struct nf_net_hook_ops’ has no member named ‘hooknum’
   .hooknum  = NF_IP_LOCAL_OUT,
    ^~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c:26:25: warning: excess elements in struct initializer
 #define NF_IP_LOCAL_OUT 3
                         ^
/var/lib/dkms/douane/0.8.2/build/douane.c:1176:15: note: in expansion of macro ‘NF_IP_LOCAL_OUT’
   .hooknum  = NF_IP_LOCAL_OUT,
               ^~~~~~~~~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c:26:25: note: (near initialization for ‘nfho_outgoing’)
 #define NF_IP_LOCAL_OUT 3
                         ^
/var/lib/dkms/douane/0.8.2/build/douane.c:1176:15: note: in expansion of macro ‘NF_IP_LOCAL_OUT’
   .hooknum  = NF_IP_LOCAL_OUT,
               ^~~~~~~~~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c:1177:4: error: ‘struct nf_net_hook_ops’ has no member named ‘pf’
   .pf       = NFPROTO_IPV4,
    ^~
/var/lib/dkms/douane/0.8.2/build/douane.c:1177:15: warning: excess elements in struct initializer
   .pf       = NFPROTO_IPV4,
               ^~~~~~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c:1177:15: note: (near initialization for ‘nfho_outgoing’)
/var/lib/dkms/douane/0.8.2/build/douane.c:1178:4: error: ‘struct nf_net_hook_ops’ has no member named ‘priority’
   .priority = NF_IP_PRI_LAST,
    ^~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c:1178:15: warning: excess elements in struct initializer
   .priority = NF_IP_PRI_LAST,
               ^~~~~~~~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c:1178:15: note: (near initialization for ‘nfho_outgoing’)
/var/lib/dkms/douane/0.8.2/build/douane.c: In function ‘initialize_module’:
/var/lib/dkms/douane/0.8.2/build/douane.c:1197:3: warning: statement with no effect [-Wunused-value]
   nf_register_net_hook//nf_register_hook(&nfho_outgoing);
   ^~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c:1200:3: error: expected ‘;’ before ‘if’
   if (initialize_activities_socket() < 0)
   ^~
/var/lib/dkms/douane/0.8.2/build/douane.c: In function ‘exit_module’:
/var/lib/dkms/douane/0.8.2/build/douane.c:1225:26: error: passing argument 1 of ‘nf_unregister_net_hook’ from incompatible pointer type [-Werror=incompatible-pointer-types]
   nf_unregister_net_hook(&nfho_outgoing);//nf_unregister_hook(&nfho_outgoing);
                          ^
In file included from /var/lib/dkms/douane/0.8.2/build/douane.c:20:0:
/usr/src/linux-headers-4.14.0-kali1-common/include/linux/netfilter.h:157:6: note: expected ‘struct net *’ but argument is of type ‘struct nf_net_hook_ops *’
 void nf_unregister_net_hook(struct net *net, const struct nf_hook_ops *ops);
      ^~~~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c:1225:3: error: too few arguments to function ‘nf_unregister_net_hook’
   nf_unregister_net_hook(&nfho_outgoing);//nf_unregister_hook(&nfho_outgoing);
   ^~~~~~~~~~~~~~~~~~~~~~
In file included from /var/lib/dkms/douane/0.8.2/build/douane.c:20:0:
/usr/src/linux-headers-4.14.0-kali1-common/include/linux/netfilter.h:157:6: note: declared here
 void nf_unregister_net_hook(struct net *net, const struct nf_hook_ops *ops);
      ^~~~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/douane/0.8.2/build/douane.c: At top level:
/var/lib/dkms/douane/0.8.2/build/douane.c:1174:31: error: storage size of ‘nfho_outgoing’ isn’t known
 static struct nf_net_hook_ops nfho_outgoing = {
                               ^~~~~~~~~~~~~
cc1: some warnings being treated as errors
/usr/src/linux-headers-4.14.0-kali1-common/scripts/Makefile.build:325: recipe for target '/var/lib/dkms/douane/0.8.2/build/douane.o' failed
make[4]: *** [/var/lib/dkms/douane/0.8.2/build/douane.o] Error 1
/usr/src/linux-headers-4.14.0-kali1-common/Makefile:1520: recipe for target '_module_/var/lib/dkms/douane/0.8.2/build' failed
make[3]: *** [_module_/var/lib/dkms/douane/0.8.2/build] Error 2
Makefile:146: recipe for target 'sub-make' failed
make[2]: *** [sub-make] Error 2
Makefile:8: recipe for target 'all' failed
make[1]: *** [all] Error 2
```

--- a/Makefile
+++ b/Makefile
@@ -42,7 +42,7 @@ MODPROBE_OUTPUT=$(shell lsmod | grep douane)
 # Compilation flags
 EXTRA_CFLAGS=-g -DDOUANE_VERSION=\"$(MODULE_VERSION)\"
 # Compilation flags with debug
-EXTRA_CFLAGS=-g -DDOUANE_VERSION=\"$(MODULE_VERSION)\" -DDEBUG
+#EXTRA_CFLAGS=-g -DDOUANE_VERSION=\"$(MODULE_VERSION)\" -DDEBUG
 
 ifeq ($(KERNEL_VERSION),)
 KERNEL_VERSION=$(shell uname -r)
@@ -50,7 +50,7 @@ endif
 
 # make
 all:
-	$(MAKE) -C /lib/modules/$(KERNEL_VERSION)/build M=$(PWD) modules
+	$(MAKE) -C /lib/modules/$(KERNEL_VERSION)/build M=$(shell pwd) modules
 
 # make clean
 clean:

--- a/Makefile
+++ b/Makefile
@@ -42,7 +42,7 @@ MODPROBE_OUTPUT=$(shell lsmod | grep douane)
 # Compilation flags
 EXTRA_CFLAGS=-g -DDOUANE_VERSION=\"$(MODULE_VERSION)\"
 # Compilation flags with debug
-# EXTRA_CFLAGS=-g -DDOUANE_VERSION=\"$(MODULE_VERSION)\" -DDEBUG
+EXTRA_CFLAGS=-g -DDOUANE_VERSION=\"$(MODULE_VERSION)\" -DDEBUG
 
 ifeq ($(KERNEL_VERSION),)
 KERNEL_VERSION=$(shell uname -r)

--- a/README.md
+++ b/README.md
@@ -4,4 +4,4 @@
 
 This repository contains the kernel module, the heart, of [the Douane firewall application](https://github.com/Douane/Douane), a Linux application firewall.
 
-Find the compilcation instructions in [the Compilation Wiki page](https://github.com/Douane/Douane/wiki/Compilation).
+Find the compilation instructions in [the Compilation Wiki page](https://github.com/Douane/Douane/wiki/Compilation).

--- a/douane.c
+++ b/douane.c
@@ -974,19 +974,6 @@ static unsigned int netfiler_packet_hook(void *priv, struct sk_buff *skb, const 
 #else
           socket_file_ino = skb->sk->sk_socket->file->f_dentry->d_inode->i_ino;
 #endif
-	  if(socket_file_ino == NULL)
-	  {
-            printk(KERN_ERR "douane:%d:%s: !!OOPS!! socket_file_ino == NULL. !!OOPS!!\n", __LINE__, __FUNCTION__);
-            kfree(activity);
-            return NF_ACCEPT;
-	  }
-
-	  if(task_info == NULL)
-	  {
-            printk(KERN_ERR "douane:%d:%s: !!OOPS!! task_info == NULL. !!OOPS!!\n", __LINE__, __FUNCTION__);
-            kfree(activity);
-            return NF_ACCEPT;
-	  }
 
           task_info = task_info_from_open_file_inode(socket_file_ino);
           if (task_info)

--- a/douane.c
+++ b/douane.c
@@ -1177,14 +1177,13 @@ static struct nf_hook_ops nfho_outgoing = {
   .pf       = NFPROTO_IPV4,
   .priority = NF_IP_PRI_LAST,
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(4,12,14)
-  .dev	    = ,
+  .dev	    = dev_get_by_name(&init_net, "enp0s25"),
 # endif
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4,4,0)
   .owner    = THIS_MODULE
 #endif
 };
 
-const struct nf_hook_state *nf_state;
 
 /*
 ** Linux kernel module initializer and cleaner methods
@@ -1193,23 +1192,18 @@ static int __init initialize_module(void)
 {
 #ifdef DEBUG
   printk(KERN_INFO "douane:%d:%s: Initializing module\n", __LINE__, __FUNCTION__);
-#endif
-  printk(KERN_INFO "douane:%d:%s: cyn TOP\n", __LINE__, __FUNCTION__);
-  /*
-  read_lock(&dev_base_lock);
-  while(dev){
-    printk(KERN_INFO "douane: cyn: found [%s]\n", dev->name);
+
+  // Net Device debugging
+  struct net_device *dev = first_net_device(&init_net);
+  while(dev)
+  {
+    printk(KERN_INFO "douane:%d:%s: Net_Device found: name: %s - ifindex: %d\n", __LINE__, __FUNCTION__, dev->name, dev->ifindex);
     dev = next_net_device(dev);
   }
-  read_unlock(&dev_base_lock);
-  */
-  printk(KERN_INFO "douane:%d:%s: cyn BOTTOM\n", __LINE__, __FUNCTION__);
-
+#endif
   INIT_LIST_HEAD(&rules.list);
   INIT_LIST_HEAD(&process_socket_inodes.list);
   
-  //printk(KERN_INFO "douane: cyn: in: %s\n", nf_state->in->name);
-  //printk(KERN_INFO "douane: cyn: out: %s\n", &nf_state->out->name);
  
   // Hook to Netfilter
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(4,12,14)

--- a/douane.c
+++ b/douane.c
@@ -17,7 +17,6 @@
 #include <linux/module.h>         // Needed by all modules
 #include <linux/kernel.h>         // Needed for KERN_INFO
 #include <linux/version.h>        // Needed for LINUX_VERSION_CODE >= KERNEL_VERSION
-#include <linux/netfilter.h>
 // ~~~~ Due to bug https://bugs.launchpad.net/ubuntu/+source/linux/+bug/929715 ~~~~
 // #undef __KERNEL__
 // #include <linux/netfilter_ipv4.h> // NF_IP_POST_ROUTING, NF_IP_PRI_LAST
@@ -27,6 +26,7 @@ enum nf_ip_hook_priorities {
   NF_IP_PRI_LAST = INT_MAX
 };
 // ~~~~
+#include <linux/netdevice.h>	  // net_device
 #include <linux/netfilter.h>      // nf_register_hook(), nf_unregister_hook(), nf_register_net_hook(), nf_unregister_net_hook()
 #include <linux/netlink.h>        // NLMSG_SPACE(), nlmsg_put(), NETLINK_CB(), NLMSG_DATA(), NLM_F_REQUEST, netlink_unicast(), netlink_kernel_release(), nlmsg_hdr(), NETLINK_USERSOCK, netlink_kernel_create()
 #include <linux/sched.h>          // for_each_process(), task_lock(), task_unlock()
@@ -1176,11 +1176,15 @@ static struct nf_hook_ops nfho_outgoing = {
   .hooknum  = NF_IP_LOCAL_OUT,
   .pf       = NFPROTO_IPV4,
   .priority = NF_IP_PRI_LAST,
+#if LINUX_VERSION_CODE <= KERNEL_VERSION(4,12,14)
+  .dev	    = ,
+# endif
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4,4,0)
   .owner    = THIS_MODULE
 #endif
 };
 
+const struct nf_hook_state *nf_state;
 
 /*
 ** Linux kernel module initializer and cleaner methods
@@ -1190,9 +1194,23 @@ static int __init initialize_module(void)
 #ifdef DEBUG
   printk(KERN_INFO "douane:%d:%s: Initializing module\n", __LINE__, __FUNCTION__);
 #endif
+  printk(KERN_INFO "douane:%d:%s: cyn TOP\n", __LINE__, __FUNCTION__);
+  /*
+  read_lock(&dev_base_lock);
+  while(dev){
+    printk(KERN_INFO "douane: cyn: found [%s]\n", dev->name);
+    dev = next_net_device(dev);
+  }
+  read_unlock(&dev_base_lock);
+  */
+  printk(KERN_INFO "douane:%d:%s: cyn BOTTOM\n", __LINE__, __FUNCTION__);
+
   INIT_LIST_HEAD(&rules.list);
   INIT_LIST_HEAD(&process_socket_inodes.list);
-
+  
+  //printk(KERN_INFO "douane: cyn: in: %s\n", nf_state->in->name);
+  //printk(KERN_INFO "douane: cyn: out: %s\n", &nf_state->out->name);
+ 
   // Hook to Netfilter
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(4,12,14)
   nf_register_hook(&nfho_outgoing);
@@ -1226,7 +1244,11 @@ static void __exit exit_module(void)
     activities_socket = NULL;
   }
 
+#if LINUX_VERSION_CODE <= KERNEL_VERSION(4,12,14)
+  nf_unregister_hook(&nfho_outgoing);
+#else
   nf_unregister_net_hook(&init_net, &nfho_outgoing);
+#endif
 
 #ifdef DEBUG
   printk(KERN_INFO "douane:%d:%s: Kernel module removed!\n", __LINE__, __FUNCTION__);

--- a/douane.c
+++ b/douane.c
@@ -27,9 +27,10 @@ enum nf_ip_hook_priorities {
   NF_IP_PRI_LAST = INT_MAX
 };
 // ~~~~
-#include <linux/netfilter.h>      // nf_register_hook(), nf_unregister_hook()
+#include <linux/netfilter.h>      // nf_register_hook(), nf_unregister_hook(), nf_register_net_hook(), nf_unregister_net_hook()
 #include <linux/netlink.h>        // NLMSG_SPACE(), nlmsg_put(), NETLINK_CB(), NLMSG_DATA(), NLM_F_REQUEST, netlink_unicast(), netlink_kernel_release(), nlmsg_hdr(), NETLINK_USERSOCK, netlink_kernel_create()
 #include <linux/sched.h>          // for_each_process(), task_lock(), task_unlock()
+#include <linux/sched/signal.h>   // for_each_process(), task_lock(), task_unlock()
 #include <linux/ip.h>             // ip_hdr()
 #include <linux/udp.h>            // udp_hdr()
 #include <linux/tcp.h>            // tcp_hdr()
@@ -1193,7 +1194,11 @@ static int __init initialize_module(void)
   INIT_LIST_HEAD(&process_socket_inodes.list);
 
   // Hook to Netfilter
+#if LINUX_VERSION_CODE <= KERNEL_VERSION(4,12,14)
   nf_register_hook(&nfho_outgoing);
+#else
+  nf_register_net_hook(&init_net, &nfho_outgoing);
+#endif
 
   // Open a Netfilter socket to communicate with the user space
   if (initialize_activities_socket() < 0)
@@ -1221,7 +1226,7 @@ static void __exit exit_module(void)
     activities_socket = NULL;
   }
 
-  nf_unregister_hook(&nfho_outgoing);
+  nf_unregister_net_hook(&init_net, &nfho_outgoing);
 
 #ifdef DEBUG
   printk(KERN_INFO "douane:%d:%s: Kernel module removed!\n", __LINE__, __FUNCTION__);

--- a/douane.c
+++ b/douane.c
@@ -1181,9 +1181,6 @@ static struct nf_hook_ops nfho_outgoing = {
   .hooknum  = NF_IP_LOCAL_OUT,
   .pf       = NFPROTO_IPV4,
   .priority = NF_IP_PRI_LAST,
-#if LINUX_VERSION_CODE <= KERNEL_VERSION(4,12,14)
-//  .dev	    = , // .dev is not used!
-# endif
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4,4,0)
   .owner    = THIS_MODULE
 #endif

--- a/douane.c
+++ b/douane.c
@@ -974,6 +974,19 @@ static unsigned int netfiler_packet_hook(void *priv, struct sk_buff *skb, const 
 #else
           socket_file_ino = skb->sk->sk_socket->file->f_dentry->d_inode->i_ino;
 #endif
+	  if(socket_file_ino == NULL)
+	  {
+            printk(KERN_ERR "douane:%d:%s: !!OOPS!! socket_file_ino == NULL. !!OOPS!!\n", __LINE__, __FUNCTION__);
+            kfree(activity);
+            return NF_ACCEPT;
+	  }
+
+	  if(task_info == NULL)
+	  {
+            printk(KERN_ERR "douane:%d:%s: !!OOPS!! task_info == NULL. !!OOPS!!\n", __LINE__, __FUNCTION__);
+            kfree(activity);
+            return NF_ACCEPT;
+	  }
 
           task_info = task_info_from_open_file_inode(socket_file_ino);
           if (task_info)
@@ -1182,7 +1195,7 @@ static struct nf_hook_ops nfho_outgoing = {
   .pf       = NFPROTO_IPV4,
   .priority = NF_IP_PRI_LAST,
 #if LINUX_VERSION_CODE <= KERNEL_VERSION(4,12,14)
-  .dev	    = dev_get_by_name(&init_net, "enp0s25"),
+//  .dev	    = , // .dev is not used!
 # endif
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4,4,0)
   .owner    = THIS_MODULE

